code/modules/power/chem_combustion.dm

```diff
@@ -0,0 +1,394 @@
+/obj/machinery/power/combustion_generator
+	name = "Portable Generator"
+	desc = "A portable combustion generator that burns fuel from a fuel tank, there is a port for a gas tank. A warning reads: DO NOT RUN INDOORS, OR WHILE UNSECURE."
+	icon_state = "chemportgen0"
+	density = 1
+	anchored = 0
+	flags = FPRINT | FLUID_SUBMERGE | NOSPLASH
+	mats = list("MET-2" = 12, "CON-1" = 8)
+	deconstruct_flags = DECON_SCREWDRIVER | DECON_WRENCH | DECON_CROWBAR | DECON_WELDER | DECON_WIRECUTTERS
+
+	var/active = 0
+	var/fuel_drain_rate = 0.3
+	var/atmos_drain_rate = 0.02
+	var/output_multiplier = 1 // for bigger generators?
+	var/last_output
+
+	var/obj/item/reagent_containers/food/drinks/fueltank/fuel_tank
+	var/obj/item/tank/inlet_tank
+
+	var/image/fuel_tank_image
+	var/image/inlet_tank_image
+
+	update_icon()
+		if (src.active)
+			if (src.last_output >= 40000)
+				src.icon_state = "chemportgen3"
+			else if (src.last_output >= 20000)
+				src.icon_state = "chemportgen2"
+			else
+				src.icon_state = "chemportgen1"
+		else
+			src.icon_state = "chemportgen0"
+
+		src.overlays = null
+		if (!src.fuel_tank_image)
+			src.fuel_tank_image = image('icons/obj/power.dmi')
+		if (!src.inlet_tank_image)
+			src.inlet_tank_image = image('icons/obj/power.dmi')
+
+		src.fuel_tank_image.icon_state = "genfueltank"
+
+		if (istype(src.inlet_tank, /obj/item/tank/oxygen) || istype(src.inlet_tank, /obj/item/tank/emergency_oxygen))
+			src.inlet_tank_image.icon_state = "gengastank_o"
+		else
+			src.inlet_tank_image.icon_state = "gengastank"
+
+		if (src.fuel_tank)
+			src.overlays += src.fuel_tank_image
+		if (src.inlet_tank)
+			src.overlays += src.inlet_tank_image
+
+		signal_event("icon_updated")
+
+	was_deconstructed_to_frame(mob/user)
+		. = ..()
+		src.stop_engine()
+
+	attack_hand(var/mob/user, params)
+		if (..(user, params))
+			return
+
+		var/dat = {"
+		<b>Status:</b><BR>
+		Engine: [src.active ? "Active" : "Stopped"] <BR>
+		Floor Bolts: [src.anchored ? "Anchored" : "Unanchored"] <BR>
+		Power Output: [src.active && src.last_output ? src.last_output : 0]W<BR><BR>
+		<b>Controls:</b><BR>
+		<A href='?src=\ref[src];engine=1'>Engine: [active ? "Stop" : "Start"]</A><BR>
+		<A href='?src=\ref[src];fuel=1'>[src.fuel_tank ? "Eject [src.fuel_tank.name]" : "Connect Fuel Tank"]</A><BR>
+		<A href='?src=\ref[src];inlet=1'>[src.inlet_tank ? "Eject [src.inlet_tank.name]" : "Connect Gas Tank"]</A><BR>
+		"}
+
+		if (user.client.tooltipHolder)
+			user.client.tooltipHolder.showClickTip(src, list(
+				"params" = params,
+				"title" = src.name,
+				"content" = dat,
+			))
+
+		return
+
+	Topic(href, href_list)
+		if (..(href, href_list))
+			return
```

Comment:
Oh my god, why would you add a new old-style window and not tgui? D:

Reply:
Had a look at making one and figured it would be a bit much for a such a small menu. Also I asked in imcoder and it seemed to be ok in tooltips.